service/history/replication/stream_receiver.go

```diff
@@ -257,26 +319,98 @@ func (r *StreamReceiverImpl) processMessages(
 		if streamResp.Err != nil {
 			return streamResp.Err
 		}
-		tasks := r.taskConverter.Convert(
+		if err := r.validateAndSetReceiverMode(streamResp.Resp.GetMessages().Priority); err != nil {
+			r.logger.Error("ReplicationTask wrong receiver mode", tag.Error(err))
+			break // sender mode changed, exit loop and let stream reconnect to retry
+		}
+
+		if err = ValidateTasksHaveSamePriority(streamResp.Resp.GetMessages().Priority, streamResp.Resp.GetMessages().ReplicationTasks...); err != nil {
+			// This should not happen because source side is sending task 1 by 1. Validate here just in case.
+			r.logger.Error("ReplicationTask priority check failed", tag.Error(err))
+			// Todo: Change to write Tasks to DLQ. As resend task will not help here
+			break // exit loop and let stream reconnect to retry task
+		}
+		convertedTasks := r.taskConverter.Convert(
 			clusterName,
 			r.clientShardKey,
 			r.serverShardKey,
 			streamResp.Resp.GetMessages().ReplicationTasks...,
 		)
 		exclusiveHighWatermark := streamResp.Resp.GetMessages().ExclusiveHighWatermark
 		exclusiveHighWatermarkTime := timestamp.TimeValue(streamResp.Resp.GetMessages().ExclusiveHighWatermarkTime)
-		r.logger.Debug(fmt.Sprintf("StreamReceiver processMessages received %d tasks with exclusiveHighWatermark %d", len(tasks), exclusiveHighWatermark))
-		for _, task := range r.taskTracker.TrackTasks(WatermarkInfo{
+		taskTracker, taskScheduler, err := r.getTrackerAndSchedulerByPriority(streamResp.Resp.GetMessages().Priority)
+		if err != nil {
+			r.logger.Error("Failed to get task tracker and scheduler", tag.Error(err))
+			// Todo: Change to write Tasks to DLQ. As resend task will not help here
+			break // exit loop and let stream reconnect to retry task
```

Comment:
return error? would stream reconnect help in this case?

Reply:
This case is very unlikely to happen. When this happens, it means there is something wrong on sender side. I changed it to reconnect stream to retry.